### DISCUSSION TITLE
Welcome to PAMI Discussions!

### DISCUSSION THREAD

**Discussion by udayRage** [UdayLab/PAMI#24] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by AlmeidaMAV** [comment]
Dear Mr. Rage,
 
I hope this email finds you well. I am writing to extend my appreciation for your exceptional work on the development of the PAMI library.
 
Allow me to introduce myself, my name is Marco Almeida, and I recently came across the PAMI library during my online research on pattern mining frameworks and libraries. After thoroughly exploring its features and capabilities, I was thoroughly impressed by the robustness, flexibility, and user-friendliness of PAMI.
 
But still I needed to write to you because after watching your YouTube videos I don’t know if I am able to use PAMI with a few databases I need to work with.
 
So, one of the data bases I would like to ask is a large two columns with an Index column and the Values column like this:

INDEX,VALUES
1,V
2,V
3,A
4,V
5,E
6,A
7,V
8,A
…
 
So, what I need is to find patterns on the VALUES that after a sequence of, let’s say,  5 VALUES the 6th VALUE is a ˜V˜.
 
Using the database above as an example, PAMI would find Indexes 2 to 7 and check if it repeats over the whole database and if it matches the support of 30% it would save to a file.
 
Or the search would be like show me sequences that match 30% support.
 
Problem is that watching the videos, all examples there have multiples values per indexes, which is different from what I have and need.
 
So, I don’t know if you understood what I want to know and my example, but if you did, could you help me with this?

Thank you for any help and best regards!!
 
Marco Almeida.

**Reply by udayRage** [reply]
Dear Marco,

Good day!

         Thank you very much for visiting our repository and posting your doubts.

         Based on your provided information, your dataset seems to be a biological/genome sequence dataset containing a sequence of characters.

        Example: VVAVEAVA ...

       Currently, PAMI does not have algorithms that can discover patterns in a biological sequence dataset. We are working on developing the algorithms; however, it will take some time.

       Please let me know if my understanding of your dataset description is incorrect. We can further discuss this here.   

Regards
Uday

**Comment by AlmeidaMAV** [comment]
Hello!!

Thank you for replying!!!

As to your understanding, I have to say that it is not correct, sorry. But I know I didn't provide too much information and therefore lead you to a  misunderstanding. Although I will explain to you what those letters mean, the end will be the same, I think.

So those letters correspond to: V = Vermelho (red), A = Azul (blue), E = Empate (draw), all the is from a virtual game in which I am saving the results and trying to find a pattern in it.

The original collected data I have also have the timestamp of each result that I didn't include when I wrote to you, but I think it doesn't help.

Well that's it.

**Reply by udayRage** [reply]
Hai,

       Thank you very much for providing more information on the data. Also, I regret my misunderstanding of your data due to a lack of communication.

       If it is the game's data, then you can model your data as a time series and discover the common patterns in multiple sequences.

     Game1: VVAVV...
     Game2: AAAXX ...
      ...

     Please check the paper titled "Discovering Periodic Patterns in Irregular Time Series"
     The dataset that I have used in this paper to explain the concepts is similar to your dataset.  

     The corresponding algorithm is available in PAMI. 

     Feel free to contact us if you have any other questions.

**Reply by AlmeidaMAV** [reply]
Ahhh ok, I will read it and will see what I can get and then report back to you here.

Thank you once again!!

Best regards,
Marco Almeida.

**Comment by AlmeidaMAV** [comment]
So, after reading the article, refactored the code so that it outputs the sequence as you suggested.

1, A V V A V A
2,V V V A A A

I also used the basics FPGrowth algo and it returned 3 results, but the output was in a way didn't understand and expected:

E:582 
V:2303 
A:2354 

That's the contents of the exported TXT file.

I also wrote a python script to search for repeating sequence throughout the database and it returned much more instances that I could visually check on the database if they matched and they did.

So, there are two things I need to figure out:

1 - Understand the way the output displays the results
2 - See if it matches with the code I have written
3 - See if the both scripts match the results, since they use the same database.

So, could you once again help me understand why/how the PAMI outputs that way?

**Reply by udayRage** [reply]
Hai, 

   I will prepare a help manual this Saturday, and share it with you. You can give it a try.

**Reply by udayRage** [reply]
Below is the code to find the patterns
==================================

from PAMI.partialPeriodicPattern.timeSeries import PPGrowth as pp
obj = pp.PPGrowth('test.txt', 3, 2, ' ')
obj.startMine()
periodicFrequentPatterns = obj.getPatterns()
print("Total number of Periodic Frequent Patterns:",␣
↪len(periodicFrequentPatterns))
obj.save("output.txt")
Df = obj.getPatternsAsDataFrame()
memUSS = obj.getMemoryUSS()
print("Total Memory in USS:", memUSS)
memRSS = obj.getMemoryRSS()
print("Total Memory in RSS", memRSS)
run = obj.getRuntime()
===============================
Sample test.txt file is

1 (a,1) (a,3) (b,2) (b,3) (c,2) (c,3)
2 (a,1) (b,1) (b,2) (b,3) (c,1)
3 (a,1) (a,2) (b,1) (b,3) (c,2)
4 (a,1) (b,1) (b,2) (c,3)
5 (a,1) (a,3) (b,3) (c,2) (c,2)
6 (a,1) (a,2) (b,2) (b,3)


Just so you know, events can appear in any order in a window. Please read the file: https://udayrage.github.io/PAMI/manuals/timeSeries.html
